[#93541780] Add LB forward of HTTPS (tcp/443) to nginx for tsuru-api

keymon opened the pull request:
**What** 
In order to [Use Tsuru API securely](https://www.pivotaltracker.com/n/projects/1275640/stories/93541780) we need to expose, on both AWS and GCE, the port 443 for the `nginx` which terminates SSL (see alphagov/tsuru-ansible#74) for `tsuru-api` servers.

**How to review it**

In this PR:
- On AWS we create the ELB configuration listening on port 443 and forwarding TCP. The health check uses HTTPS:443/info, which tests both `nginx` and `tsuru-api` service.
- On GCE:
  - We explicitly allocate a new ip address, to use it in two different Google compute forwarding rules.
  - we add a second `google_compute_forwarding_rule` for HTTPS/443, but uses the healthcheck in plain HTTP, with HTTP:8080/info. The reason is that the HTTPS check is not supported in terraform and only in the GCE beta API.
- we did NOT delete the old 8080 configuration. This will be done in the next story.

tsuru clients will fail until we deploy a valid certificate, so you can test the SSL termination with curl:

```
curl https://<envname>-api.tsuru2.paas.alphagov.co.uk -k
```

**Who should review this**

@actionjack and @keymon worked on this. @dcarley did the ansible nginx related work.

**Dependencies**
- It is required to merge/use the PR alphagov/tsuru-ansible#74 to deploy nginx for SSL termination
- The certificate is invalid, so the tsuru clients will fail and refuse to connect.

actionjack committed:
Make ELB's forward tcp/443 versus terminate SSL

We don't currently have SSL on the API in GCE environments because the
load balancers which do SSL termination aren't currently supported by
Terraform.

We do have ELBs terminating SSL in AWS, but the certificate upload
process is outside of Terraform.

In order to harmonise how we deal with SSL termination across both
platforms (`gce` and `aws`) and ensure we can fully automate the
generation of environments we will forward the TCP traffic to a
[Nginx](http://wiki.nginx.org/Main) `SSL` terminator on the `tsuru api server`.

This commit consists of:

* [Changing the AWS load balancer to forward to TCP:443 instead of terminating SSL](https://github.com/alphagov/tsuru-ansible/pull/74)
* Changing the healthcheck to check on `HTTPS:443/info` of the `tsuru api server`

keymon committed:
Add a dedicated address to the api LB for gce

If you don't specify an `ip_address` when creating a
`google_compute_forwarding_rule` it will use a ephemeral ip.
We need to allocate an ip by declaring it with a
`google_compute_address` in order to use it in different
forwarding rules, like the HTTPS one.

keymon committed:
Add the HTTPS load balancing for api in gce

We need to setup a new `google_compute_forwarding_rule` in order
to provide HTTPS load balancing for the api running on gce.

In this commit we:

* Created a new check, as we need to specify a different port: 443
* Setup a new pool, as we need to use a different check
* Create the forwarding rule, referencing the new pool. It reuses
  the `google_compute_address` assigned to the 8080 rule.

We added a `https` suffix to all the resources, as done in the
`router` resources, but we keep the names for the old
resources to avoid conflicts applying terraform.

dcarley commented on gce/api-servers.tf:
```diff
@@ -27,13 +27,37 @@ resource "google_compute_http_health_check" "api" {
 }
+resource "google_compute_http_health_check" "api_https" {
+  name = "${var.env}-tsuru-api-https"
+  port = 443
```
Does this work? I'd assumed that pointing an HTTP health check at an HTTPS endpoint would fail, because it would try to talk HTTP to something that wanted to speak TLS.

dcarley commented on gce/api-servers.tf:
```diff
@@ -27,13 +27,37 @@ resource "google_compute_http_health_check" "api" {
 resource "google_compute_target_pool" "api" {
   name = "${var.env}-tsuru-api-lb"
   instances = [ "${google_compute_instance.api.*.self_link}" ]
   health_checks = [ "${google_compute_http_health_check.api.name}" ]
```
I think it'd be preferable to have a single (and same type of) health check for both pools. But I guess this'll get tidied up as part of disabling HTTP:8080?

actionjack replied:
Yes we are using the expansion contraction pattern to add the https and then remove the http to port 8080 later.

dcarley commented on aws/api-servers.tf:
```diff
@@ -31,11 +31,10 @@ resource "aws_elb" "api-ext" {
+    instance_protocol = "tcp"
     lb_port = 443
-    lb_protocol = "https"
-    ssl_certificate_id = "${var.api_ssl_certificate_id}"
```
We can remove `api_ssl_certificate_id` from `variables.tf` too.

keymon replied:
You are right, done.

keymon committed:
Remove the still unsupported gce HTTPS healthcheck

Currently HTTPS health checks in GCE are part of the GCE beta API and are not
supported by terraform.

There's a [feature request](https://github.com/hashicorp/terraform/issues/2190)
where these possible implementations are discussed:

 1. Alternate resource type: google_compute_https_health_check
 2. Add a new attribute `protocol`

In this commit remove the HTTPS check and use only HTTP on port 8080, for
both the `HTTP/8080` and `HTTPS/443` forwarding rules. The drawbacks are:

 * We don't check the nginx SSL terminator service in the health check.
 * The insecure API must be open, at least for the LB layer.

keymon commented:
We remove the still unsupported gce HTTPS healthcheck,  in the commit remove the HTTPS check and use only HTTP on port 8080, for both the HTTP/8080 and HTTPS/443 forwarding rules. The drawbacks are:

 * We don't test the nginx SSL terminator service in the health check
 * The insecure API must be open, at least for the LB layer.

keymon committed:
Remove the legacy api_ssl_certificate_id var

As we don't terminate SSL in ELB, we don't need the certificate
reference in `variables.tf`

dcarley commented on aws/api-servers.tf:
```diff
@@ -31,11 +31,10 @@ resource "aws_elb" "api-ext" {
-    instance_port = 8080
-    instance_protocol = "http"
+    instance_port = 443
+    instance_protocol = "tcp"
```
Should we add an additional `listener` to do HTTP _and_ HTTPS? That would match what we're doing for GCE and we could cleanup both as part of the "shut down insecure Tsuru API" story.

Sorry, I missed this the first time.

actionjack replied:
I propose that we shouldn't since:
-  It feels a little redundant to implement a temporary workaround for GCE which we will fix in time 
- We are adding unnecessary configuration in AWS so it matches a temporary deficiency in GCE. 

Though if you feel strongly about it we can add it in.

dcarley replied:
I misread. We're keeping the 8080 listener.
